### ISSUE TITLE
Add interface to allow Spam filtering by third-party apps

### ISSUE THREAD

**Issue by cketti** [k9mail/k-9#1508]
I don't want to include a Spam filter into the app itself. Instead we could add an interface that allows third-party apps to provide Spam filtering capabilities.

**Comment by mikeloeven** [comment]
I dislike this because it is reliant on additional apps and resources adding a basic spam filter to the app is something that should have been done by default and this seems like a bit of a copout to avoid adding what is a base feature of most email clients

**Comment by jandechent** [comment]
There seems to have been a previous attemt to just include very simple filter https://github.com/Yogu/k-9 i don't know of the quality of this implementation, but it might be worth a shot?

**Comment by LongtimeUserFS** [comment]
I am a longtime user of K9-mail in Germany. 
But unfortunately I have no skills in programming at all. On my search for a solution to the exact same problem that mikeloeven described allready in #1490 I am really disappointed that there seam to be no progress in dealing with this issue. I also since many years have a provider called 1blu that doesn`t filter obvious SPAM emails. That was no problem with my PC for many years. Just since I use my smartphone with K9-Mail which still doesn`t filter obvious SPAM and also doesn`t allow me to filter on my own it became really bad. Consequently I get a lot of SPAM Mails daily and can`t change it.

I would really appreciate if anyone who has the skills to do so would try to find a solution in programming something to solve this issue. I totally agree with mikeloevens argumentation in #1490 that a spam-filter is a basic service of any good and competitive email-provider.

With regards from Germany

**Comment by wwp** [comment]
I second this feature request. Not all provider filter spam, some only add a header that indicates a score (my provider does it, but fortunately I don't rely on their scoring, and with my several email accounts, I get from 2 to 40% spam, spending half of my K-9 time to delete emails)!
Adding a way to 1) add a bayesian spam filter (surprisingly, such engines in Android seem simply.. absent), 2) allow a third-party spam filter app or 3) filter incoming messages, would be a great value added to this app.

**Comment by Biswajee** [comment]
I would like to work upon this issue as described in my proposal. 
Regards,
Biswajit Roy.

**Comment by 9662** [comment]
>  a spam-filter is a basic service of any good and competitive email-provider.

But K-9 is an email user agent, not an email provider. Providers are in general in a much better position to accurately detect and classify spam than clients (though endpoint classification also has its advantages). In any event, a spam detection and classification engine is not a trivial project and development-wise does not have much in common with writing an email user agent, so it does make sense for filtering functionality to be implemented via an external application through a defined API, in the same way as other important but non-core functions are done, such as address books, encryption, attachment display, etc.

**Comment by myset** [comment]
Maybe give a try,
It would be useful _already classified emails_ "Subject [SPAM], Email Headers Spam-assasin...." to be filtered some how, because all "[SPAM]" messages are present in INBOX
Maybe some simplified rules like Mozilla Thunderbird

Sample Classified Email Headers
> X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on mail.myset.ro
X-Spam-Flag: YES
X-Spam-Level: ************
X-Spam-Status: Yes, score=12.7 required=4.0 tests=BAYES_99,BAYES_999,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HTML_IMAGE_ONLY_16,HTML_MESSAGE,
	HTML_SHORT_LINK_IMG_2,RDNS_NONE,SPF_HELO_NONE,SPF_PASS,URIBL_ABUSE_SURBL,
	URIBL_BLACK,URIBL_DBL_SPAM,URIBL_SBL,URIBL_SBL_A autolearn=no
	autolearn_force=no version=3.4.0

**Comment by ueen** [comment]
I agree, my email provider do not seem to filter anything and i constantly classify emails as spam, often they are completly identical and completly obviously spam, maybe something like this might work https://github.com/ptnplanet/Java-Naive-Bayes-Classifier
I tried to do a fork, but the project structure seems way to complicated for my skillset, hope someone gets around to add a basic filter...

**Comment by mehdisadeghi** [comment]
FWIW, a sane provider will implement something on top of [sieve](http://sieve.info/) and [rspamd](rspamd.com/) to mark mails as spam when you move them to the spam folder.

**Comment by ueen** [comment]
Yea you don't choose your provider you know.
I did some search and FairEmail actually seems to have some kinda of Bayesian classification, I'll see if it works :)
